File: libs/monobitmap-manager/src/main/kotlin/mono/bitmap/manager/factory/PointChar.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.bitmap.manager.factory

import kotlin.math.abs

/**
 * A class representing a char at a point for generating bitmap.
 */
class PointChar private constructor(val left: Int, val top: Int, val char: Char) {

    companion object {
        fun point(left: Int, top: Int, char: Char): Sequence<PointChar> =
            sequenceOf(PointChar(left, top, char))

        fun horizontalLine(
            beginExclusive: Int,
            endExclusive: Int,
            top: Int,
            char: Char
        ): Sequence<PointChar> {
            if (abs(beginExclusive - endExclusive) <= 1) {
                return emptySequence()
            }
            val delta = if (beginExclusive < endExclusive) 1 else -1
            val begin = beginExclusive + delta
            val end = endExclusive - delta
            val range = if (begin > end) begin downTo end else begin..end
            return range.asSequence().map { PointChar(it, top, char) }
        }

        fun verticalLine(
            left: Int,
            beginExclusive: Int,
            endExclusive: Int,
            char: Char
        ): Sequence<PointChar> {
            if (abs(beginExclusive - endExclusive) <= 1) {
                return emptySequence()
            }
            val delta = if (beginExclusive < endExclusive) 1 else -1
            val begin = beginExclusive + delta
            val end = endExclusive - delta
            val range = if (begin > end) begin downTo end else begin..end
            return range.asSequence().map { PointChar(left, it, char) }
        }
    }
}

File: monosketch-svelte/src/lib/mono/monobitmap/manager/point-char.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { range } from "$libs/sequence";

/**
 * A class representing a char at a point for generating bitmap.
 */
export class PointChar {
    constructor(public left: number, public top: number, public char: string) {
    }

    static point(left: number, top: number, char: string): PointChar[] {
        return [new PointChar(left, top, char)];
    }

    static horizontalLine(beginExclusive: number, endExclusive: number, top: number, char: string): PointChar[] {
        if (Math.abs(beginExclusive - endExclusive) <= 1) {
            return [];
        }
        const delta = beginExclusive < endExclusive ? 1 : -1;
        const begin = beginExclusive + delta;
        return Array.from(range(begin, endExclusive, delta), (left, _) => new PointChar(left, top, char));
    }

    static verticalLine(left: number, beginExclusive: number, endExclusive: number, char: string): PointChar[] {
        if (Math.abs(beginExclusive - endExclusive) <= 1) {
            return [];
        }
        const delta = beginExclusive < endExclusive ? 1 : -1;
        const begin = beginExclusive + delta;
        return Array.from(range(begin, endExclusive, delta), (top, _) => new PointChar(left, top, char));
    }
}
